torch/distributions/transformed_distribution.py

```diff
@@ -0,0 +1,77 @@
+import torch
+from numbers import Number
+from torch.autograd import Variable
+from torch.distributions.bijectors import Bijector
+from torch.distributions.distribution import Distribution
+
+
+class TransformedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies a sequence of Bijectors to a base distribution.
+    Let f be the composition of bijectors applied,
+    X ~ BaseDistribution
+    Y = f(X) ~ TransformedDistribution(BaseDistribution, f)
+    log p(Y) = log p(X) + log det (dX/dY)
+    """
+    def __init__(self, base_distribution, bijectors=[], *args, **kwargs):
+        super(TransformedDistribution, self).__init__(*args, **kwargs)
+        self.has_rsample = base_distribution.has_rsample
+        self.base_dist = base_distribution
+        if isinstance(bijectors, Bijector):
+            self.bijectors = [bijectors, ]
+        elif isinstance(bijectors, list):
+            for bijector in bijectors:
+                if not isinstance(bijector, Bijector):
+                    raise ValueError("bijectors must be a Bijector or a list of Bijectors")
+            self.bijectors = bijectors
+
+    def sample(self, sample_shape=torch.Size()):
+        """
+        Generates a sample_shape shaped sample or sample_shape shaped batch of
+        samples if the distribution parameters are batched. Samples first from base distribution
+        and applies bijector.forward() for every bijector in the list.
+        """
+        x = self.base_dist.sample(sample_shape)

```

Comment:
I feel like the logic in both of these could be simplified by keeping the intermediate cache stuff internal to `Bijector` What do you think?
```py
x = self.base_dist.sample(sample_shape)
for b in self.bijectors:
    x = b.forward(x)
return x
```